### PR TITLE
Bug 1723869: Terminate custom resource watches when storage is destroyed

### PR DESCRIPTION
**Pull request by p0lyn0mial**
backports kubernetes/kubernetes#77816 and kubernetes/kubernetes#78029

### CODE REVIEW

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -96,6 +99,9 @@ type crdHandler struct {
 	masterCount int
 
 	converterFactory *conversion.CRConverterFactory
+
+	// request timeout we should delay storage teardown for
+	requestTimeout time.Duration
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: this wasn't clean

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 2]
```diff
@@ -229,7 +243,7 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		string(types.MergePatchType),
 	}
 
-	var handler http.HandlerFunc
+	var handlerFunc http.HandlerFunc
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: block from 246 to 267 as well.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/test/integration/change_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
+/*
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/test/integration/change_test.go]: had to add that file manually.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 4]
```diff
@@ -438,3 +438,39 @@ func TestWatcherNotGoingBackInTime(t *testing.T) {
 		}
 	}
 }
+
+func TestCacheWatcherStoppedOnDestroy(t *testing.T) {
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go]: I had to add that function manually.

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 5]
```diff
@@ -134,7 +142,8 @@ func NewCustomResourceDefinitionHandler(
 	establishingController *establish.EstablishingController,
 	serviceResolver webhook.ServiceResolver,
 	authResolverWrapper webhook.AuthenticationInfoResolverWrapper,
-	masterCount int) (*crdHandler, error) {
+	masterCount int,
+	requestTimeout time.Duration) (*crdHandler, error) {
```
Comment (p0lyn0mial) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: this wasn't clean as well.

**Comment by sttts** [comment]
/lgtm
/approve

@derekwaynecarr anything to do to get that into the 4.1.z channel?

**Comment by eparis** [comment]
I don't see pointers to the master PRs (or the 4.2 BZ).  this is not going to make 4.1.4,   but if we can get the 4.2 BZ VERIFIED it will be re-reviewed for 4.1.5 next week.

**Comment by sttts** [comment]
@eparis here is the master PR https://github.com/openshift/origin/pull/23248

@smarterclayton you explicitly were asking for getting this in ASAP. It probably plays a role in upgrades when CRDs change.

**Comment by jwforres** [comment]
@sttts I don't see where a 4.2 BZ was opened for QE to verify this fixes the bug, if you can get that opened today and put it into MODIFIED then QE may get to verify it in 4.2 tonight so we can let this backport into 4.1.5 tmrw

**Comment by stevekuznetsov** [comment]
Bugzilla server gave us a 200 but didn't add the external link :|

```
{"client":"bugzilla","component":"hook","file":"prow/bugzilla/client.go:112","func":"k8s.io/test-infra/prow/bugzilla.(*client).request","id":1723869,"level":"debug","method":"AddExternalBug","msg":"Got response from Bugzilla.","num":23267,"org":"openshift","repo":"origin","response":200,"time":"2019-07-01T23:26:35Z"}
```

**Pull request merged**